Allow cost center in entry of balance sheet accounts

sanjay-hercules opened the pull request:
This feature will facilitate user to maintain balance sheet accounts at cost center level.

Future Enhancement:
* Allow user to maintain warehouse at cost center level

sanjay-hercules committed:
Allow Cost Center In Entry of Balance Sheet Account

sanjay-hercules committed:
Merge branch 'develop' into cc_change_v1.0

sanjay-hercules committed:
Add parent cost center in get payment entry

sanjay-hercules committed:
Add Tests for Allow Cost Center In Entry of Balance Sheet Account

sanjay-hercules committed:
Merge branch 'develop' into cc_change_v1.0

sanjay-hercules committed:
Add tests for cost center wise account and party balance

sanjay-hercules committed:
Merge branch 'develop' into cc_change_v1.0

sanjay-hercules commented:
@nabinhait , please review and let me know if any change is required.

sanjay-hercules committed:
Merge branch 'develop' into cc_change_v1.0

nabinhait commented on erpnext/accounts/doctype/payment_entry/payment_entry.js:
```diff
@@ -317,7 +318,8 @@ frappe.ui.form.on('Payment Entry', {
 					"account": account,
-					"date": frm.doc.posting_date
+					"date": frm.doc.posting_date,
+					"cost_center": frm.doc.cost_center
```
I don't think party outstanding balance should be fetched based on selected cost center. Party has nothing to do with your cost center. So, mostly they will pay against total outstanding amount.

sanjay-hercules replied:
In ideal scenario each branch has its own bank account and maintains branch wise receivable and payable except the case where a company has centralized finance and purchase.
So if company A is having head office in Mumbai and branches in Delhi, Kolkatta and Chennai. Then in this case the branch receivable and payable will settled by branch itself not by Head Office.

Tagging Cost Center in Balance Sheet accounts is similar to maintaining branch wise financial books (e.g.. Trial Balance, P&L and Balance Sheet).

jk2081 replied:
I think we would require both pieces of information of cc-wise outstanding and total outstanding. We have cases where customers pay based on total as Navin has commented.

nabinhait replied:
Ok, as cost center is optional here, we can fetch outstanding balance as per selected cost center. If user wants to process against total outstanding, then they will leave cost center blank.

nabinhait commented on erpnext/accounts/doctype/payment_entry/payment_entry.js:
```diff
@@ -859,3 +861,10 @@ frappe.ui.form.on('Payment Entry Deduction', {
 		frm.events.set_unallocated_amount(frm);
 	}
 })
+frappe.ui.form.on('Payment Entry', {
```
As per my above comment, if balance comes as per total outstanding, this piece of code is not needed, I think. Also  this will cause asynchronous issue, as there are multiple ajax call involved.

sanjay-hercules replied:
>As per my above comment, if balance comes as per total outstanding, this piece of code is not needed, I think. 

This is required to show party balance of selected cost center. If no cost center is selected, then show overall balance at company level.

>Also this will cause asynchronous issue, as there are multiple ajax call involved.

Agreed. Will enhance this.

jk2081 replied:
Again, I think if cost centre is selected, perhaps show cost centre and total for entire company.

nabinhait commented on erpnext/accounts/doctype/purchase_invoice/purchase_invoice.js:
```diff
@@ -520,4 +520,12 @@ frappe.ui.form.on("Purchase Invoice", {
+})
+frappe.ui.form.on('Purchase Invoice', {
+	cost_center: function(frm, cdt, cdn) {
+		erpnext.utils.copy_parent_value_in_all_row(frm.doc, cdt, cdn, "items", "cost_center", "cost_center");
```
Why copying parent cost center to item rows? Expenses can be booked against multiple cost center and it should be based on item's default cost center.

sanjay-hercules replied:
The purpose of copying parent cost center to item is to book purchase against cost center selected in parent document to maintain cost center wise payable.

jk2081 replied:
Once again, my comment here of being able to see the total payable even when viewing CC wise accounts. The idea is that we would CC for product lines similar to classes in QBO so same branch may have more than one CCs. This is the scenario why visibility of both CC-level and total outstanding data would be useful.

nabinhait replied:
Let's don't copy parent CC to child items. Keep them separate. We have a plan to introduce "Department" and "Branch" inside item table, so that we can get P&L analysis based on branch / department.

sanjay-hercules replied:
> Let's don't copy parent CC to child items. Keep them separate. We have a plan to introduce "Department" and "Branch" inside item table, so that we can get P&L analysis based on branch / department.

Okay, I will remove function to copy parent CC to child.

nabinhait commented on erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py:
```diff
@@ -658,7 +665,7 @@ def make_write_off_gl_entry(self, gl_entries):
 					"credit_in_account_currency": self.base_write_off_amount \
 						if write_off_account_currency==self.company_currency else self.write_off_amount,
-					"cost_center": self.write_off_cost_center
+					"cost_center": self.cost_center or self.write_off_cost_center
```
priority should be given to `self.write_off_cost_center`

sanjay-hercules replied:
If user opted to tag cost center in Balance Sheet Accounts then write-off amount should also go to Cost Center selected in parent document otherwise trial balance wouldn't match.

nabinhait replied:
ok

nabinhait commented on erpnext/accounts/doctype/purchase_invoice/purchase_invoice.py:
```diff
@@ -673,7 +680,7 @@ def make_gle_for_rounding_adjustment(self, gl_entries):
 					"debit_in_account_currency": self.rounding_adjustment,
 					"debit": self.base_rounding_adjustment,
-					"cost_center": round_off_cost_center,
+					"cost_center": self.cost_center or round_off_cost_center,
```
priority should be given to `round_off_cost_center `

sanjay-hercules replied:
If user opted to tag cost center in Balance Sheet Accounts then rounded-off amount should also go to Cost Center selected in parent document otherwise trial balance wouldn't match.

nabinhait replied:
ok

nabinhait commented on erpnext/accounts/doctype/sales_invoice/sales_invoice.js:
```diff
@@ -783,3 +783,11 @@ var select_loyalty_program = function(frm, loyalty_programs) {
+frappe.ui.form.on('Sales Invoice', {
+	cost_center: function(frm, cdt, cdn) {
+		erpnext.utils.copy_parent_value_in_all_row(frm.doc, cdt, cdn, "items", "cost_center", "cost_center");
+		erpnext.utils.copy_parent_value_in_all_row(frm.doc, cdt, cdn, "taxes", "cost_center", "cost_center");
```
same as purchase invoice comment?

sanjay-hercules replied:
The purpose of copying parent cost center to item is to book sales against cost center selected in parent document to maintain cost center wise receivable.

nabinhait replied:
as per purchase invoice

sanjay-hercules replied:
Okay, I will remove function to copy parent CC to child.

nabinhait commented on erpnext/accounts/doctype/sales_invoice/sales_invoice.py:
```diff
@@ -754,13 +755,14 @@ def make_loyalty_point_redemption_gle(self, gl_entries):
 				self.get_gl_dict({
 					"account": self.loyalty_redemption_account,
-					"cost_center": self.loyalty_redemption_cost_center,
+					"cost_center": self.cost_center or self.loyalty_redemption_cost_center,
```
priority should be given to `loyalty_redemption_cost_center`

sanjay-hercules replied:
If user opted to tag cost center in Balance Sheet Accounts then loyalty_redemption amount should also go to Cost Center selected in parent document otherwise trial balance wouldn't match.

nabinhait replied:
ok

nabinhait commented on erpnext/accounts/doctype/sales_invoice/sales_invoice.py:
```diff
@@ -852,7 +859,7 @@ def make_write_off_gl_entry(self, gl_entries):
 					"debit_in_account_currency": self.base_write_off_amount \
 						if write_off_account_currency==self.company_currency else self.write_off_amount,
-					"cost_center": self.write_off_cost_center or default_cost_center
+					"cost_center": self.cost_center or self.write_off_cost_center or default_cost_center
```
`self.write_off_cost_center or or self.cost_center or default_cost_center`

sanjay-hercules replied:
If user opted to tag cost center in Balance Sheet Accounts then write-off amount should also go to Cost Center selected in parent document otherwise trial balance wouldn't match.

nabinhait replied:
ok

nabinhait commented on erpnext/accounts/doctype/sales_invoice/sales_invoice.py:
```diff
@@ -867,7 +874,7 @@ def make_gle_for_rounding_adjustment(self, gl_entries):
 					"credit_in_account_currency": self.rounding_adjustment,
 					"credit": self.base_rounding_adjustment,
-					"cost_center": round_off_cost_center,
+					"cost_center": self.cost_center or round_off_cost_center,
```
`round_off_cost_center or self.cost_center`

sanjay-hercules replied:
If user opted to tag cost center in Balance Sheet Accounts then round_off amount should also go to Cost Center selected in parent document otherwise trial balance wouldn't match.

nabinhait replied:
ok

nabinhait reviewed (changes requested):
- Rebase your PR please

FinForceConsulting commented:
@sanjay-hercules "this feature will facilitate user to maintain balance sheet accounts at cost center level"

What is the Use Case of showing Balance Sheet items on Cost Center Level? Cost Center or Profit Center is used to slice and dice the P&L situation and generally, BalanceSheet is looked into the whole as of a particular day.

Splitting Payable and Receivable at a Cost Center level is not quite right unless there is a separate entity of a Customer or a Supplier is dealing with a particular Cost Center.

Regards
CA Liyakat

agritheory commented:
> What is the Use Case of showing Balance Sheet items on Cost Center Level?

@FinForceConsulting Liyakat, this allows you to calculate an ROI or ROA per cost center, which is very useful if you have cost centers set up as physical locations within one business. There's a less strong case for it if you're measuring departments or a product line, but it's still good to have receivables and payables in there so you can get an accrual profitability out of your cost center.

FinForceConsulting commented:
@Tyler, do you have a sample report analysed in that way to calculate ROI. And it used BS items for it. I would love to see that.

See, I know QBO does this split of BS items onto Class (kind of cost center) but I have never used it.

_Liyakat

sanjay-hercules committed:
Merge branch 'develop' into cc_change_v1.0

sanjay-hercules committed:
set parent cost center in taxes

sanjay-hercules committed:
Merge branch 'develop' into cc_change_v1.0

jk2081 commented:
A general comment is that even one branch may have multiple cost centers for ex. based on product lines. And the cost centers may share vendors and customers. In this scenario, it is important to view the total amounts and cost center level amounts of either payables or receivables since customers do pay based on total outstanding regardless of our definition of cost center.

sanjay-hercules committed:
Merge branch 'develop' into cc_change_v1.0

sanjay-hercules committed:
1. Remove copy parent cost_center to child
2. Improve update party and account balance functionality on cost_center change
3. Add cost_center filter to get_outstanding_documents

sanjay-hercules committed:
Merge branch 'develop' into cc_change_v1.0

sanjay-hercules commented:
@nabinhait Done below changes:
1. Remove copy parent cost_center to child from Sales and Purchase Invoice
2. Improve update party and account balance functionality on change of cost_center in Payment Entry
3. Add cost_center filter to get_outstanding_documents in payment entry
Please review.

nabinhait commented on erpnext/accounts/doctype/payment_entry/payment_entry.js:
```diff
@@ -859,3 +862,38 @@ frappe.ui.form.on('Payment Entry Deduction', {
+								frm.set_value("party_balance", r.message.party_balance);
+							},
+							() => {
+								if(frm.doc.payment_type != "Internal") {
```
Internal Transfer

sanjay-hercules replied:
Getting outstanding documents is not required for Internal  Transfer.

sanjay-hercules replied:
Rebase and push code.

nabinhait reviewed (changes requested):
Please rebase and push your code. We have fixed unrelated test cases which were failing on travis.

sanjay-hercules committed:
Merge branch 'develop' into cc_change_v1.0

sanjay-hercules committed:
fix Codacy and Travis issue

sanjay-hercules committed:
Merge branch 'develop' into cc_change_v1.0

sanjay-hercules committed:
Merge branch 'develop' into cc_change_v1.0

sanjay-hercules committed:
Merge branch 'develop' into cc_change_v1.0

jk2081 commented:
Looking forward to this. @sanjay-hercules and @nabinhait thank you for your work. Not a small PR.
